### PR TITLE
Combinators for handling failures

### PR DESCRIPTION
**Pull request by michaelneale**
@etorreborre @charleso @domdere How does this sit here? 

I am still on the fence if useful.

### CODE REVIEW

File: x-either/src/X/Control/Monad/Trans/Either.hs [hunk 1]
```diff
@@ -46,3 +51,26 @@ joinErrors g1 g2 = mapEitherT (fmap (join . bimap g2 (first g1))) . runEitherT
...
+--
+runOrFail :: (Monad m) => EitherT Text m a -> m a
+runOrFail = (=<<) (either (fail . unpack) return) . runEitherT
+
+-- |
+-- Useful to check an EitherT is Right, eg:
+--   expectRight (awsErrorRender) $ runAWST ...
+--
+expectRight :: Functor m => (l -> Text) -> EitherT l m a -> EitherT Text m a
+expectRight err e = bimapEitherT (err) id $ e
```
Comment (markhibberd) [inline x-either/src/X/Control/Monad/Trans/Either.hs]: Think this is just `firstEitherT`

File: x-either/src/X/Control/Monad/Trans/Either.hs [hunk 2]
```diff
@@ -46,3 +51,26 @@ joinErrors g1 g2 = mapEitherT (fmap (join . bimap g2 (first g1))) . runEitherT
...
 joinErrorsEither = joinErrors Right Left
+
+
+-- |
+-- Great for testing. Allows you to keep all properties in EitherT all the way to the top.
+-- Great for testIO:
+-- prop_... = testIO . runOrFail $ do <all EitherT here>
+--
+runOrFail :: (Monad m) => EitherT Text m a -> m a
+runOrFail = (=<<) (either (fail . unpack) return) . runEitherT
```
Comment (domdere) [inline x-either/src/X/Control/Monad/Trans/Either.hs]: very similar to [`orDie`] (https://github.com/ambiata/x/blob/master/x-optparse/src/X/Options/Applicative.hs#L74-L75)

I dont know if i see the value of it as the testing scenario you describe is very similar to

``` Haskell
foo :: (Monad m) => EitherT Text m Property -> m Property
foo = eitherT (return . failWith) return
```

which i dont really have a problem with typing

maybe you can change this to:

``` Haskell
unifyLeft :: (e -> m a) -> EitherT e m a -> m a
unifyLeft = flip eitherT return
```

Comment (domdere) [inline x-either/src/X/Control/Monad/Trans/Either.hs]: then when testing you can compactly get your fail prop out with `unifyLeft failWith`.  my main concern with this function is ppl misusing it to drop errors silently in proper prod apps with `unifyLeft (const ())`

File: x-either/src/X/Control/Monad/Trans/Either.hs (hunk 1 above)
Comment (michaelneale) [inline x-either/src/X/Control/Monad/Trans/Either.hs]: yes, it is. I just tried. *sigh*.

File: x-either/src/X/Control/Monad/Trans/Either.hs (hunk 2 above)
Comment (michaelneale) [inline x-either/src/X/Control/Monad/Trans/Either.hs]: yeah you don't want fail in prod code (orDie is also only really for careful use too). Yeah this doesn't seem appropriate here.

Comment (etorreborre) [inline x-either/src/X/Control/Monad/Trans/Either.hs]: Why not create it in `x-either-test`? Too much burden for one function?

Comment (michaelneale) [inline x-either/src/X/Control/Monad/Trans/Either.hs]: unifyLeft looks handy, but it is one of those unusual things that you wouldn't realise exists until you have re-invented it anyway I suspect.

Comment (michaelneale) [inline x-either/src/X/Control/Monad/Trans/Either.hs]: where is x-either-test?

Comment (etorreborre) [inline x-either/src/X/Control/Monad/Trans/Either.hs]: That's the burden: you have to create a new module :-)

Comment (michaelneale) [inline x-either/src/X/Control/Monad/Trans/Either.hs]: ah gotcha. Wasn't sure if I was missing something.

**Comment by michaelneale** [comment]
yeah I think the cost/benefit is out of wack for this. Not obvious enough to add code/deps to pretty core libraries of re-use. Closing it for now.

**Pull request closed without merging**